File: neetcode150/medium/853-Car-Fleet.py
class Solution:
  def carFleet(self, target: int, position: list, speed: list) -> int:
    cars = []
    for i in range(len(position)):
      cars.append((position[i], (target - position[i]) / speed[i]))
    cars = sorted(cars, key = lambda x: x[0])

    stack = []

    for car in cars[::-1]:
      if len(stack) == 0:
        stack.append(car)
      elif car[1] > stack[-1][1]:
        stack.append(car)

    return len(stack)

File: neetcode150/medium/cpp/853-Car-Fleet.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        int N = speed.size();
        if (N == 1) return 1;

        vector<pair<int, double>> ps(N); 
        for (int i = 0; i < N; i++){
            ps[i] = {position[i], speed[i]};
        }

        sort(ps.begin(), ps.end());

        double curr, time;
        int res = 0;
        int i = N-1;
        time = (target - ps[i].first) / ps[i].second;
        for (int j = N-2; j >= 0; j--) {
            curr = (target - ps[j].first) / ps[j].second;
            if (curr <= time) {
                curr = min(time, curr);
            } else {
                time = curr;
                ++res;
            }
        }
        
        return res + 1;
    }
};